File: CardPeak.UI.ClientApp/src/services/api/agentsController.ts
import axios from 'axios'

const API = {
    GET_ALL: '/agents',
    GET_AGENT_DASHBOARD: (agentId: number) => {
        return '/agents/' + agentId;
    },
    GET_AGENT_SAVINGS: (agentId: number) => {
        return '/agents/' + agentId + '/savings'
    },
    GET_AGENT_SAVINGS_BY_YEAR: (agentId: number, year: number) => {
        return '/agents/' + agentId + '/savings?year=' + year
    },
    GET_AGENT_DASHBOARD_FILTER: (agentId: number) => {
        return '/agents/' + agentId + "/filter"
    },
    DEBIT_AGENT: (agentId: number) => {
        return '/agents/' + agentId + '/debit'
    },
    CREDIT_AGENT: (agentId: number) => {
        return '/agents/' + agentId + '/credit'
    },
    INCENTIVE_AGENT: (agentId: number) => {
        return '/agents/' + agentId + '/incentive'
    },
    SAVINGS_CREDIT_AGENT: (agentId: number) => {
        return '/agents/' + agentId + '/savings/credit'
    },
    SAVINGS_DEBIT_AGENT: (agentId: number) => {
        return '/agents/' + agentId + '/savings/debit'
    },
    UPDATE_AGENT: (agentId: number) => {
        return '/agents/' + agentId + '/update'
    },
    CREATE_AGENT: '/agents/create',
    GET_DETAILS: (agentId: number) => {
        return '/agents/' + agentId + '/details';
    },
    GET_AGENT_PAYOUT: '/agents/payout',
    DEACTIVATE_AGENT: (agentId: number) => {
        return "/agents/deactivate/" + agentId
    }
}

export function deactivateAgent(agentId: number, successCallback: () => void, errorCallback?: (e: string) => void) {
    axios.post(API.DEACTIVATE_AGENT(agentId))
        .then((r) => {
            successCallback();
        })
        .catch((reason) => {
            if (errorCallback) {
                errorCallback(reason.message);
            }
        })
}

export function getAgentPayout(successCallback: (data: CardPeak.Models.AgentPayoutModel) => void, errorCallback: (e: string) => void) {
    axios.get(API.GET_AGENT_PAYOUT)
        .then((r) => {
            successCallback(r.data as CardPeak.Models.AgentPayoutModel);
        })
        .catch((reason) => {
            errorCallback(reason.message);
        });
}

export function getAll(successCallback: (data: CardPeak.Entities.Agent[]) => void) {
    axios.get(API.GET_ALL)
        .then((r) => {
            successCallback(r.data as CardPeak.Entities.Agent[]);
        });
}

export function getAgentDashboard(agentId: number,
    successCallback: (data: CardPeak.Entities.AgentDashboard) => void) {

    axios.get(API.GET_AGENT_DASHBOARD(agentId))
        .then((r) => {
            successCallback(r.data as CardPeak.Entities.AgentDashboard);
        });
}

export function getAgentSavings(agentId: number,
    successCallback: (data: CardPeak.Entities.AgentSavings) => void) {

    axios.get(API.GET_AGENT_SAVINGS(agentId))
        .then((r) => {
            successCallback(r.data as CardPeak.Entities.AgentSavings);
        });
}

export function getAgentSavingsByYear(agentId: number, year: number,
    successCallback: (data: CardPeak.Entities.AgentSavings) => void) {

    axios.get(API.GET_AGENT_SAVINGS_BY_YEAR(agentId, year))
        .then((r) => {
            successCallback(r.data as CardPeak.Entities.AgentSavings);
        });
}

export function getAgentDashboardFiltered(agentId: number, startDate?: string, endDate?: string,
    successCallback?: (data: CardPeak.Entities.AgentDashboard) => void) {

    axios.get(API.GET_AGENT_DASHBOARD_FILTER(agentId), ({
            params: {
                startDate: startDate,
                endDate: endDate
            }
        }))
        .then((r) => {
            if (successCallback) {
                successCallback(r.data as CardPeak.Entities.AgentDashboard);
            }
        });
}

export function postAgentTransaction(
    transaction: CardPeak.Entities.DebitCreditTransaction,
    transactionType: Transaction,
    successCallback: (data: CardPeak.Entities.DebitCreditTransaction) => void,
    errorCallback: (error: string) => void) {

    let url: string;
    switch (transactionType.toLowerCase()) {
        case "credit":
            url = API.CREDIT_AGENT(transaction.agentId);
            break;
        case "debit":
            url = API.DEBIT_AGENT(transaction.agentId);
            break;
        case "incentive":
            url = API.INCENTIVE_AGENT(transaction.agentId);
            break;
        case "savings-credit":
            url = API.SAVINGS_CREDIT_AGENT(transaction.agentId);
            break;
        case "savings-debit":
            url = API.SAVINGS_DEBIT_AGENT(transaction.agentId);
            break;
    }

    axios.post(url, null, {
            params: {
                amount: transaction.amount, remarks: transaction.remarks, transactionDate: transaction.transactionDateTime
            }
        })
        .then((r) => {
            successCallback(r.data as CardPeak.Entities.DebitCreditTransaction);
        })
        .catch((reason) => {
            errorCallback(reason.message);
        });
}

export function updateAgent(agent: CardPeak.Entities.Agent,
    successCallback: (data: CardPeak.Entities.Agent) => void,
    errorCallback: (error: string) => void) {

    let url = API.UPDATE_AGENT(agent.agentId);
    axios.post(url, { ...agent })
        .then((r) => {
            successCallback(r.data as CardPeak.Entities.Agent);
        })
        .catch((reason) => {
            errorCallback(reason.message);
        })
}

export function createAgent(agent: CardPeak.Entities.Agent,
    successCallback: (data: CardPeak.Entities.Agent) => void,
    errorCallback: (error: string) => void) {

    let url = API.CREATE_AGENT;
    axios.post(url, { ...agent })
        .then((r) => {
            successCallback(r.data as CardPeak.Entities.Agent);
        })
        .catch((reason) => {
            errorCallback(reason.message);
        })
}

export async function getDetailsAsync(agentId: number) {
    let result = await axios(API.GET_DETAILS(agentId));
    if (result.status != 200) {
        return undefined;
    }
    return result.data as CardPeak.Entities.AgentDetails;
}

File: CardPeak.WebAPI/Controllers/AgentsController.cs
using CardPeak.Core.Service;
using CardPeak.Domain;
using CardPeak.Service;
using CardPeak.WebAPI.Core;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using System.Web;
using System.Web.Http;

namespace CardPeak.WebAPI.Controllers
{
    [RoutePrefix("api/agents")]
    public sealed class AgentsController : ApiController
    {
        private static readonly string Root = HttpContext.Current.Server.MapPath("~/App_Data/profiles");
        private static readonly string Extension = ".png";
        private static readonly string ContentType = "image/png";

        private IAgentService AgentService { get; set; }

        public AgentsController()
        {
            this.AgentService = new AgentService(new Repository.EF.CardPeakDbContext());
        }

        protected override void Dispose(bool disposing)
        {
            base.Dispose(disposing);
            if (disposing)
            {
                this.AgentService.Dispose();
            }
        }

        public IEnumerable<Agent> GetAllAgents()
        {
            return this.AgentService.GetAllAgents();
        }

        public IHttpActionResult GetAgent(int id)
        {
            return this.GetAgent(id, null, null);
        }

        [HttpGet]
        [Route("{id}/details")]
        public IHttpActionResult GetDetails(int id)
        {
            var result = this.AgentService.GetAgentDetails(id);
            if (result == null)
            {
                return this.NotFound();
            }

            return this.Ok(result);
        }

        [HttpGet]
        [Route("{id}/filter")]
        public IHttpActionResult GetAgent(int id, [FromUri] DateTime? startDate = null, [FromUri] DateTime? endDate = null)
        {
            var result = this.AgentService.GetAgentDashboard(
                id,
                startDate ?? new DateTime(DateTime.Today.Year, DateTime.Today.Month, 1),
                endDate);

            if (result == null)
            {
                return this.NotFound();
            }

            return this.Ok(result);
        }

        [HttpPost]
        [Route("{id}/update")]
        public IHttpActionResult Update(Agent agent)
        {
            var result = this.AgentService.Update(agent);
            if (result == null)
            {
                return this.InternalServerError();
            }

            return this.Ok(result);
        }

        [HttpGet]
        [Route("payout")]
        public IHttpActionResult GetPayout()
        {
            try
            {
                var result = this.AgentService.GetAgentPayouts();
                return this.Ok(result);
            }
            catch (Exception e)
            {
                return this.InternalServerError(e);
            }
        }

        [HttpPost]
        [Route("create")]
        public IHttpActionResult Create(Agent agent)
        {
            var result = this.AgentService.Create(agent);
            if (result == null)
            {
                return this.InternalServerError();
            }

            return this.Ok(result);
        }

        [HttpPost]
        [Route("{id}/credit")]
        public IHttpActionResult CreditAgent(int id, decimal amount, string remarks)
        {
            return this.DebitCreditAgent(id, amount, remarks, false);
        }

        [HttpPost]
        [Route("{id}/debit")]
        public IHttpActionResult DebitAgent(int id, decimal amount, string remarks)
        {
            return this.DebitCreditAgent(id, amount, remarks, true);
        }

        private IHttpActionResult DebitCreditAgent(int id, decimal amount, string remarks, bool isDebit)
        {
            var result = this.AgentService.AddDebitCreditTransaction(id, amount, remarks, isDebit);
            if (result == null)
            {
                return this.InternalServerError();
            }

            return this.Ok(result);
        }

        [HttpPost]
        [Route("{id}/incentive")]
        public IHttpActionResult IncentiveAgent(int id, decimal amount, string remarks)
        {
            var result = this.AgentService.AddIncentiveTransaction(id, amount, remarks);
            if (result == null)
            {
                return this.InternalServerError();
            }

            return this.Ok(result);
        }

        [HttpGet]
        [Route("{id}/savings")]
        public IHttpActionResult GetAgentSavings(int id, [FromUri] int? year = null)
        {
            var result = this.AgentService.GetAgentSavings(id, year);
            if (result == null)
            {
                return this.InternalServerError();
            }

            return this.Ok(result);
        }

        [HttpPost]
        [Route("{id}/savings/credit")]
        public IHttpActionResult SavingsCreditAgent(int id, decimal amount, string remarks, DateTime? transactionDate)
        {
            return this.SavingsDebitCreditAgent(id, amount, remarks, transactionDate, false);
        }

        [HttpPost]
        [Route("{id}/savings/debit")]
        public IHttpActionResult SavingsDebitAgent(int id, decimal amount, string remarks, DateTime? transactionDate)
        {
            return this.SavingsDebitCreditAgent(id, amount, remarks, transactionDate, true);
        }


        private IHttpActionResult SavingsDebitCreditAgent(int id, decimal amount, string remarks, DateTime? transactionDate, bool isDebit)
        {
            var result = this.AgentService.AddDebitCreditSavingsTransaction(id, amount, remarks, transactionDate, isDebit);
            if (result == null)
            {
                return this.InternalServerError();
            }

            return this.Ok(result);
        }

        [HttpPost]
        [Route("deactivate/{id}")]
        public IHttpActionResult DeactivateAgent(int id)
        {
            try
            {
                this.AgentService.DeactivateAgent(id);
                return this.Ok(id);
            }
            catch (Exception e)
            {
                return this.InternalServerError(e);
            }
        }

        [HttpPost]
        [Route("{id}/photo")]
        public async Task<IHttpActionResult> UploadPhoto(int id)
        {
            HttpRequestMessage request = this.Request;
            if (!request.Content.IsMimeMultipartContent())
            {
                throw new HttpResponseException(HttpStatusCode.UnsupportedMediaType);
            }

            try
            {
                var provider = new AgentProfileMultipartFormDataStreamProvider(AgentsController.Root, id, AgentsController.Extension);
                await Request.Content.ReadAsMultipartAsync(provider);
                var uploadedFile = new FileInfo(provider.FileData.First().LocalFileName);

                return this.Ok();
            }
            catch (System.Exception e)
            {
                return this.InternalServerError(e);
            }
        }

        [HttpGet]
        [Route("{id}/photo")]
        public IHttpActionResult GetPhoto(int id)
        {
            var filePath = string.Format("{0}\\{1}{2}", AgentsController.Root, id, AgentsController.Extension);

            if (!File.Exists(filePath))
            {
                return NotFound();
            }

            byte[] imageBytes = File.ReadAllBytes(filePath);
            HttpResponseMessage response = new HttpResponseMessage(HttpStatusCode.OK);
            response.Content = new ByteArrayContent(imageBytes);
            response.Content.Headers.ContentType = new MediaTypeHeaderValue(AgentsController.ContentType);

            return ResponseMessage(response);
        }
    }
}
